source/common/config/watch_map.cc

```diff
@@ -0,0 +1,171 @@
+#include "common/config/watch_map.h"
+
+namespace Envoy {
+namespace Config {
+
+Watch* WatchMap::addWatch(SubscriptionCallbacks& callbacks) {
+  auto watch = std::make_unique<Watch>(callbacks);
+  Watch* watch_ptr = watch.get();
+  wildcard_watches_.insert(watch_ptr);
+  watches_.insert(std::move(watch));
+  return watch_ptr;
+}
+
+void WatchMap::removeWatch(Watch* watch) {
+  wildcard_watches_.erase(watch); // may or may not be in there, but we want it gone.
+  watches_.erase(watch);
+}
+
+AddedRemoved WatchMap::updateWatchInterest(Watch* watch,
+                                           const std::set<std::string>& update_to_these_names) {
+  if (update_to_these_names.empty()) {
+    wildcard_watches_.insert(watch);
+  } else {
+    wildcard_watches_.erase(watch);
+  }
+
+  std::vector<std::string> newly_added_to_watch;
+  std::set_difference(update_to_these_names.begin(), update_to_these_names.end(),
+                      watch->resource_names_.begin(), watch->resource_names_.end(),
+                      std::inserter(newly_added_to_watch, newly_added_to_watch.begin()));
+
+  std::vector<std::string> newly_removed_from_watch;
+  std::set_difference(watch->resource_names_.begin(), watch->resource_names_.end(),
+                      update_to_these_names.begin(), update_to_these_names.end(),
+                      std::inserter(newly_removed_from_watch, newly_removed_from_watch.begin()));
+
+  watch->resource_names_ = update_to_these_names;
+
+  return AddedRemoved(findAdditions(newly_added_to_watch, watch),
+                      findRemovals(newly_removed_from_watch, watch));
+}
+
+absl::flat_hash_set<Watch*> WatchMap::watchesInterestedIn(const std::string& resource_name) {
+  // Note that std::set_union needs sorted sets. Better to do it ourselves with insert().
```

Comment:
Could we just use an ordered set? `std::less` gives a total ordering of pointers in C++14.

Reply:
We could, but then updates would be more expensive; watch_interest_ would need to hold ordered sets. Removed that comment; it actually took me a minute to figure out what it was even saying!